File: Engine/src/main/java/org/etieskrill/engine/entity/service/PointShadowMappingService.java
package org.etieskrill.engine.entity.service;

import org.etieskrill.engine.entity.Entity;
import org.etieskrill.engine.entity.component.Drawable;
import org.etieskrill.engine.entity.component.PointLightComponent;
import org.etieskrill.engine.entity.data.Transform;
import org.etieskrill.engine.entity.data.TransformC;
import org.etieskrill.engine.graphics.Renderer;
import org.etieskrill.engine.graphics.gl.shader.Shaders;

import java.util.List;
import java.util.Objects;

import static org.lwjgl.opengl.GL11C.GL_DEPTH_BUFFER_BIT;
import static org.lwjgl.opengl.GL11C.glClear;

public class PointShadowMappingService implements Service {

    private final Renderer renderer;
    private final Shaders.DepthCubeMapArrayShader shader;

    public PointShadowMappingService(Renderer renderer, Shaders.DepthCubeMapArrayShader shader) {
        this.renderer = renderer;
        this.shader = shader;
    }

    @Override
    public boolean canProcess(Entity entity) {
        return entity.hasComponents(PointLightComponent.class);
    }

    @Override
    public void preProcess(List<Entity> entities) {
        entities.stream()
                .map(entity -> entity.getComponent(PointLightComponent.class))
                .filter(Objects::nonNull)
                .map(PointLightComponent::getShadowMap)
                .filter(Objects::nonNull)
                .distinct()
                .forEach(shadowMap -> {
                    shadowMap.bind();
                    glClear(GL_DEPTH_BUFFER_BIT);
                    shadowMap.unbind();
                });
    }

    @Override
    public void process(Entity targetEntity, List<Entity> entities, double delta) {
        PointLightComponent component = targetEntity.getComponent(PointLightComponent.class);
        if (component.getShadowMap() == null) return; //TODO resolve by using separate component for shadow maps

        shader.setLight(component.getLight());
        shader.setIndex(component.getShadowMapIndex());
        shader.setShadowCombined(component.getCombinedMatrices());
        shader.setFarPlane(component.getFarPlane());

        component.getShadowMap().bind();
        for (int i = 0; i < entities.size(); i++) {
            if (i == targetEntity.getId()) continue;

            Transform transform = entities.get(i).getComponent(Transform.class);
            Drawable drawable = entities.get(i).getComponent(Drawable.class);
            if (transform == null || drawable == null) continue;

            TransformC finalTransform = new Transform(transform).compose(drawable.getModel().getInitialTransform());
            renderer.render(finalTransform, drawable.getModel(), shader, null);
        }
        component.getShadowMap().unbind();
    }

}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/entity/service/impl/PointShadowMappingService.kt
package org.etieskrill.engine.entity.service.impl

import org.etieskrill.engine.entity.Entity
import org.etieskrill.engine.entity.component.Drawable
import org.etieskrill.engine.entity.component.PointLightComponent
import org.etieskrill.engine.entity.component.Transform
import org.etieskrill.engine.entity.service.Service
import org.etieskrill.engine.graphics.animation.Animator
import org.etieskrill.engine.graphics.pipeline.Pipeline
import org.etieskrill.engine.graphics.pipeline.PipelineConfig
import org.etieskrill.engine.graphics.renderer.Renderer
import org.etieskrill.engine.graphics.shader.impl.DepthCubeMapArrayShader
import org.joml.Matrix4f
import org.joml.Matrix4fc

class PointShadowMappingService(
    private val renderer: Renderer,
    private val shader: DepthCubeMapArrayShader
) : Service {

    private val DUMMY_MATRIX: Matrix4fc = Matrix4f()

    private val updateFrequency = 2
    private var cycle = 0

    override fun canProcess(entity: Entity) =
        entity.hasComponents<PointLightComponent>()
                && entity.getComponent<PointLightComponent>()!!.shadowMap != null

    override fun preProcess(delta: Double, entities: List<Entity>) {
        if (++cycle >= updateFrequency) {
            cycle = 0
        } else return

        entities.mapNotNull { it.getComponent<PointLightComponent>() }
            .mapNotNull { it.shadowMap }
            .distinct()
            .forEach { it.clear() }
    }

    override fun process(targetEntity: Entity, entities: List<Entity>, delta: Double) {
        if (cycle++ != 0) return

        val component = targetEntity.getComponent<PointLightComponent>()!!

        shader.apply {
            TODO("fix")
//            light = component.light
//            index = component.shadowMapIndex!!
//            shadowCombined = component.shadowCombinedMatrices!! as Array<Matrix4fc>
//            farPlane = component.shadowFarPlane!!
        }

        entities.filterNot { it.id == targetEntity.id }
            .forEach { entity ->
                val transform = entity.getComponent<Transform>() ?: return@forEach
                val drawable = entity.getComponent<Drawable>() ?: return@forEach

                entity.getComponent<Animator>()?.let { //TODO animated shader
//                    shader.boneMatrices = it.transformMatricesArray
                }
//                shader.combined = DUMMY_MATRIX

                drawable.model.nodes.forEach { node ->
//                    shader.model = (transform * node.getGlobalTransform()).matrix

                    node.meshes.forEach { mesh ->
                        val pipeline = Pipeline(
                            mesh.vao,
                            PipelineConfig(),
                            shader,
                            component.shadowMap!!
                        )

                        renderer.render(pipeline)
                    }
                }
            }
    }

}
